### DISCUSSION TITLE
How do you rewrite urls?

### DISCUSSION THREAD

**Discussion by graphixela** [wintercms/winter#1173] [General]
I'm used to doing this with htaccess but I'm on an nginx server.
I tried multiple things and nothing worked. I'm guessing because WinterCMS has it's own rewrite logic.
I'm stumped as to what I'm supposed to do.

I have /agent?name=variable and I'm trying to get the variable to be the subdomain and stay on the same page.
So /agent?name=graphixela becomes graphixela.domain.com

    location ~ ^/agent?name=([^/]+)(.*)$ {
        rewrite $1.example.com$2;
    }
The last few I had were
 rewrite agent?name=$1 ^agent/([0-9]+)/?$ break;
 rewrite ^/agent$ /agent/?name=$1 last; 
and I know those don't do the subdomain but I was trying to get something simpler to work. I also tried with index.php before the Winter rewrite was called. I'm confused as to which URL variable is supposed to go first as I've seen both in examples that are supposed to do the same thing.
I don't know if I need to use try_files before either.
Apache was so much easier. Any help would be appreciated.

**Comment by LukeTowers** [comment]
Please post your nginx config. This is more of an nginx question than a Winter CMS question, although you could perhaps try using the [Winter.Redirect](https://github.com/wintercms/wn-redirect-plugin) if you would like.

**Reply by graphixela** [reply]
I tried redirect and it worked but it brought the subdomain back to the homepage which isn't correct,

Here is the config 

```
server {
    listen      %ip%:%web_port%;
    server_name %domain_idn% %alias_idn%;
    root        %docroot%;
    index       index.php index.html index.htm;
    access_log  /var/log/nginx/domains/%domain%.log combined;
    access_log  /var/log/nginx/domains/%domain%.bytes bytes;
    error_log   /var/log/nginx/domains/%domain%.error.log error;

    include %home%/%user%/conf/web/%domain%/nginx.forcessl.conf*;

    location = /favicon.ico {
        log_not_found off;
        access_log off;
    }

    location = /robots.txt {
        allow all;
        log_not_found off;
        access_log off;
    }

    location = /locationlist.json {
        allow all;
        log_not_found off;
        access_log off;
    }

    location = /markers.xml {
        allow all;
        log_not_found off;
        access_log off;
    }

    location / {
        rewrite ^/agent$ /agent/index.php?name=$1 last;
        rewrite ^/.*$ /index.php last;
        try_files $uri $uri/ /index.php?$args;
        location ~* ^.+\.(ogg|ogv|svg|svgz|swf|eot|otf|woff|woff2|mov|mp3|mp4|w>
            expires 30d;
            fastcgi_hide_header "Set-Cookie";
        }

         location ~ [^/]\.php(/|$) {
            fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
            try_files $uri =404;
            fastcgi_pass %backend_lsnr%;
            fastcgi_index index.php;
            include /etc/nginx/fastcgi_params;
            include %home%/%user%/conf/web/%domain%/nginx.fastcgi_cache.conf*;

        }
    }

    # Whitelist
    ## Let October handle if static file not exists
    location ~ ^/favicon\.ico { try_files $uri /index.php; }
    location ~ ^/sitemap\.xml { try_files $uri /index.php; }
    location ~ ^/robots\.txt { try_files $uri /index.php; }
    location ~ ^/humans\.txt { try_files $uri /index.php; }
    location ~ ^/markers\.xml { try_files $uri /index.php; }
    location ~ ^/locationlist\.json { try_files $uri /index.php; }

    ## Let nginx return 404 if static file not exists
    location ~ ^/storage/app/uploads/public { try_files $uri 404; }
    location ~ ^/storage/app/media { try_files $uri 404; }
    location ~ ^/storage/temp/public { try_files $uri 404; }

    location ~ ^/modules/.*/assets { try_files $uri 404; }
    location ~ ^/modules/.*/resources { try_files $uri 404; }
    location ~ ^/modules/.*/behaviors/.*/assets { try_files $uri 404; }
    location ~ ^/modules/.*/behaviors/.*/resources { try_files $uri 404; }
    location ~ ^/modules/.*/widgets/.*/assets { try_files $uri 404; }
    location ~ ^/modules/.*/widgets/.*/resources { try_files $uri 404; }
    location ~ ^/modules/.*/formwidgets/.*/assets { try_files $uri 404; }
    location ~ ^/modules/.*/formwidgets/.*/resources { try_files $uri 404; }
    location ~ ^/modules/.*/reportwidgets/.*/assets { try_files $uri 404; }
    location ~ ^/modules/.*/reportwidgets/.*/resources { try_files $uri 404; }

    location ~ ^/plugins/.*/.*/assets { try_files $uri 404; }
    location ~ ^/plugins/.*/.*/resources { try_files $uri 404; }
    location ~ ^/plugins/.*/.*/behaviors/.*/assets { try_files $uri 404; }
    location ~ ^/plugins/.*/.*/behaviors/.*/resources { try_files $uri 404; }
    location ~ ^/plugins/.*/.*/reportwidgets/.*/assets { try_files $uri 404; }
    location ~ ^/plugins/.*/.*/reportwidgets/.*/resources { try_files $uri 404;>
    location ~ ^/plugins/.*/.*/formwidgets/.*/assets { try_files $uri 404; }
    location ~ ^/plugins/.*/.*/formwidgets/.*/resources { try_files $uri 404; }
    location ~ ^/plugins/.*/.*/widgets/.*/assets { try_files $uri 404; }
    location ~ ^/plugins/.*/.*/widgets/.*/resources { try_files $uri 404; }

    location ~ ^/themes/.*/assets { try_files $uri 404; }
    location ~ ^/themes/.*/resources { try_files $uri 404; }

    error_page  403 /error/404.html;
    error_page  404 /error/404.html;
    error_page  500 502 503 504 /error/50x.html;


    location /error/ {
        alias   %home%/%user%/web/%domain%/document_errors/;
    }

    location /vstats/ {
        alias   %home%/%user%/web/%domain%/stats/;
        include %home%/%user%/web/%domain%/stats/auth.conf*;
    }

    include /etc/nginx/conf.d/phpmyadmin.inc*;
    include /etc/nginx/conf.d/phppgadmin.inc*;
    include %home%/%user%/conf/web/%domain%/nginx.conf_*;

}
```

**Comment by bennothommo** [comment]
@graphixela My understanding of your question is that you want the following redirects to occur:

- `domain.com/agent?name=foo` to redirect to `foo.domain.com`
- `domain.com/agent?name=bar` to redirect to `bar.domain.com`

And that any part of the URL after `?name=` is tacked on to the end of the redirected URL - ie.

- `domain.com/agent?name=foo/next/page` to redirect to `foo.domain.com/next/page`

I'm not quite sure that's possible. You are using a query string for the `name` variable, which by conventions of RFC 3986 can only be followed by a hashbang. You can't follow it up with more path sections.

You can most certainly do these rewrites though:

- `domain.com/agent/foo/next/page` to redirect to `foo.domain.com/next/page`
- `domain.com/agent?name=foo&path=next/page` to redirect to `foo.domain.com/next/page`

**Reply by graphixela** [reply]
You're combining the two. The second example I used was just to demonstrate it wasn't working.
I just need the first part:

domain.com/agent?name=foo to rewrite to foo.domain.com

But it needs to stay on the agent page. Basically it's a mask/rewrite instead of a redirect.

I know I can do something. My issue is nothing is working and I don't know why that is.
As you can see in my config file I had tried to save markers.xml to the root directory and it didn't work even thought it should. I ended up just saving it to assets.

Changing the URL did work with the redirect plugin but it goes to the homepage instead of staying on the current agent page.
Seeing as how the plugin does redirects and not rewrites, I don't think the plugin would work for my case. Unless I'm missing something.

**Reply by bennothommo** [reply]
Right, so what you're saying is that `domain.com/agent?name=foo` redirects to `foo.domain.com`, and you would like `foo.domain.com` to actually mean (according to the PHP side) the original `domain.com/agent?name=foo`.

It's gonna be difficult, but it is possible.

You're going to need to be able to extract the subdomain portion first from the `server_name`, which means that you'll probably want this in its own `server` block, and will likely need to copy the entire config you have there now so that Winter and assets still work correctly.

```
server_name  ~^(?<subdomain>.+)\.domain\.com$;
```

The above extracts the subdomain portion to a variable `$subdomain`. Then you'd need to dynamically set the `REQUEST_URI` parameter in FastCGI. Something like:

```
fastcgi_param  REQUEST_URI /agent?name=$subdomain
```

This needs to occur *after* the following line:

```
include /etc/nginx/fastcgi_params;
```

Since by default, Nginx will simply set `REQUEST_URI` to be the request URI sent in the browser.

Given all the placeholders in your config, I'm assuming you're using a webhosting platform, so you may need to be careful with this, as I would assume that platform will likely rewrite or modify the config and may override your changes. Hopefully however, this should give you enough to go on.

Good luck!

**Reply by graphixela** [reply]
Thanks Ben. I appreciate it.

I have my own VPS so I'll just make another template.
